core/common/src/SourceExt.kt

```diff
@@ -0,0 +1,318 @@
+ *
+ * The scan terminates at either [endIndex] or source's exhaustion, whichever comes first. The
+ * maximum number of bytes scanned is `toIndex-fromIndex`.
+ * If [b] not found in buffered data, [endIndex] is yet to be reached and the underlying source is not yet exhausted
+ * then new data will be read from the underlying source into the buffer.
+ *
+ * @param b the value to find.
+ * @param startIndex the start of the range (inclusive) to find [b], `0` by default.
+ * @param endIndex the end of the range (exclusive) to find [b], [Long.MAX_VALUE] by default.
+ *
+ * @throws IllegalStateException when the source is closed.
+ * @throws IllegalArgumentException when `startIndex > endIndex` or either of indices is negative.
+ */
+public fun Source.indexOf(b: Byte, startIndex: Long = 0L, endIndex: Long = Long.MAX_VALUE): Long {
+    require(startIndex in 0..endIndex) { "startIndex: $startIndex, endIndex: $endIndex" }
+    if (startIndex == endIndex) return -1L
+
+    var offset = startIndex
+    val peekSource = peek()
+
+    if (!peekSource.request(offset)) {
+        return -1L
+    }
+    peekSource.skip(offset)
+    while (offset < endIndex && peekSource.request(1)) {
+        if (peekSource.readByte() == b) return offset
+        offset++
+    }
+    return -1L
+}
+
+/**
+ * Removes all bytes from this source and returns them as a byte array.
+ *
+ * @throws IllegalStateException when the source is closed.
+ */
+public fun Source.readByteArray(): ByteArray {
+    return readByteArrayImpl( -1)
+}
+
+/**
+ * Removes [byteCount] bytes from this source and returns them as a byte array.
+ *
+ * @param byteCount the number of bytes that should be read from the source.
+ *
+ * @throws IllegalArgumentException when [byteCount] is negative.
+ * @throws EOFException when the underlying source is exhausted before [byteCount] bytes of data could be read.
+ * @throws IllegalStateException when the source is closed.
+ */
+public fun Source.readByteArray(byteCount: Int): ByteArray {
+    require(byteCount >= 0) { "byteCount: $byteCount" }
+    return readByteArrayImpl(byteCount)
+}
+
+@OptIn(InternalIoApi::class)
+private fun Source.readByteArrayImpl(size: Int): ByteArray {
+    var arraySize = size
+    if (size == -1) {
+        var fetchSize = Int.MAX_VALUE.toLong()
+        while (buffer.size < Int.MAX_VALUE && request(fetchSize)) {
+            fetchSize *= 2
+        }
+        check (buffer.size < Int.MAX_VALUE) { "Can't create an array of size ${buffer.size}" }
+        arraySize = buffer.size.toInt()
+    } else {
+        require(size.toLong())
+    }
+    val array = ByteArray(arraySize)
+    buffer.readTo(array)
+    return array
+}
+
+
+/**
+ * Removes exactly `sink.length` bytes from this source and copies them into [sink].
+ *
+ * @throws EOFException when the requested number of bytes cannot be read.
+ * @throws IllegalStateException when the source is closed.
+ */
+public fun Source.readTo(sink: ByteArray) {
```

Comment:
is there a reason to not to have `start/end` indices in this declaration?

Reply:
No, there isn't. I'll add indices, thanks for pointing to it.